lang/src/main/kotlin/org/partiql/lang/eval/ExprValueExtensions.kt

```diff
@@ -363,23 +367,19 @@ fun ExprValue.cast(
                 }
                 it.toLong()
             }
-            valueFactory.newInt(result)
+            ExprValue.newInt(result)
         }
-        is FloatType -> valueFactory.newFloat(this.toDouble())
+        is FloatType -> ExprValue.newFloat(this.toDouble())
         is DecimalType -> {
             if (this.isNaN || this.isNegInf || this.isPosInf) {
                 castFailedErr("Can't convert Infinity or NaN to DECIMAL.", internal = false)
             }
 
             when (typedOpBehavior) {
-                TypedOpBehavior.LEGACY -> valueFactory.newFromIonValue(
-                    this.coerce(BigDecimal::class.java).ionValue(valueFactory.ion)
-                )
+                TypedOpBehavior.LEGACY -> ExprValue.newDecimal(coerce(BigDecimal::class.java))
                 TypedOpBehavior.HONOR_PARAMETERS ->
                     when (type.precisionScaleConstraint) {
-                        DecimalType.PrecisionScaleConstraint.Unconstrained -> valueFactory.newFromIonValue(
-                            this.coerce(BigDecimal::class.java).ionValue(valueFactory.ion)
-                        )
+                        DecimalType.PrecisionScaleConstraint.Unconstrained -> ExprValue.newDecimal(coerce(BigDecimal::class.java))
                         is DecimalType.PrecisionScaleConstraint.Constrained -> {
```

Comment:
I guess this is because `Unconstrained` is an object, https://github.com/partiql/partiql-lang-kotlin/blob/main/lang/src/main/kotlin/org/partiql/lang/types/StaticType.kt#L442, while `Constrained` is a class.